DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+qualifiers are optional.
+
+### Semantics
+
+This sections discusses all the aspects regarding the semantic analysis and interaction with other
+components of the copy constructor
+
+#### Copy constructor usage
+
+The copy constructor is implicitly used by the compiler whenever a struct variable is initialized
+as a copy of another variable of the same unqualified type:
+
+1. When a variable is explicitly initialized:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+void main()
+{
+    A a;
+    A b = a; // copy constructor gets called
+    b = a;   // assignment, not initialization
+}
+```
+
+2. When a parameter is passed by value to a function:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+void fun(A a) {}
+
+void main()
+{
+    A a;
+    fun(a);    // copy constructor gets called
+}
+```
+
+3. When a parameter is returned by value from a function:
+
+```d
+struct A
+{
+    @implicit this(ref A another) {}
+}
+
+A fun()
+{
+    A a;
+    return a;
+}
+
+void main()
+{
+    A a;
+    a = fun();      // copy constructor gets called
+}
+```
+
+The parameter of the copy constructor is passed by reference, so initializations are going
+to be lowered to copy constructor calls only if the source is an lvalue. Although this can be
+worked around by declaring temporary lvalues which can be forwarded to the copy constructor,
+binding rvalues to lvalues is beyond the scope of this DIP.
+
+Note that when a function returns a struct that defines a copy constructor, the copy constructor
+is called on the callee site, before actually returning:
+
+```d
+struct A
+{
+    @implicit this(ref A rhs) {}
+}
+
+A fun()
+{
+    A a;
+    return a;      // lowered to return tmp.copyCtor(a)
+    // return A();  // rvalue, no copyCtor called
+}
+
+void main()
+{
+    A b = fun();    // the return value of fun() is moved to the location of b
+}
+```
+
+#### Requirements
+
+1. The type of the parameter to the copy constructor needs to be identical to `typeof(this)`; an error is issued
+otherwise:
+
+```d
+struct A
+{
+    int a;
+    string b;
+}
+
+struct C
+{
+    A a;
+    string b;
+    @implicit this(ref C another)       // ok, typeof(this) == C
+    {
+        this.a = another.a;
+        this.b = another.b;
+    }
+
+    @implicit this(ref A another)       // error typeof(this) != C
+    {
+        this.a = another;
+        this.b = "hello";
+    }
+}
+
+void main()
+{
+    C c;
+    C d = c;   // ok
+    A a;
+    C e = a;   // error
+}
+```
+2. It is illegal to declare a copy constructor for a struct that has a postblit defined and vice versa:
+
+```d
+struct A
+{
+    this(this) {}
+    @implicit this(ref A another) {}        // error, struct A defines a postblit
+}
+```
+
+Note that structs that do not define a postblit explicitly but contain fields that
+define one have a generated postblit. From the copy constructor perspective it makes
+no difference whether the postblit is user defined or generated:
+
```

Comment:
I think this will not compose very well with structs that contain legacy fields that have a postblit. Can't the implicitly generated postblit just be omitted instead?

What about the reverse: a struct that has a postblit, but also fields with copy constructor?

Reply:
It's the same, postblit and copy constructor just can't go together. I've thought about preferring copy constructor over postblit, but the postblitt is essentially broken and must be deprecated. The code that defines a postblit can simply transform it into a copy constructor my changing the declaration.

Reply:
It is ok that postblit and copy constructor don't coexist in a struct, but fields could still just have a postblit that is used when the copy constructor of the outer struct initializes the field.